src/plugins/vis_augmenter/public/utils/utils.ts

```diff
@@ -0,0 +1,182 @@
+    const ast = buildExpression(visLayerExpressionFns).toAst();
+    return formatExpression(ast);
+  } catch (e) {
+    throw new Error('Expression function from augment-vis saved objects could not be generated');
+  }
+};
+
+/**
+ * Returns an error with an aggregated message about all of the
+ * errors found in the set of VisLayers. If no errors, returns undefined.
+ */
+export const getAnyErrors = (visLayers: VisLayer[], visTitle: string): Error | undefined => {
+  const visLayersWithErrors = visLayers.filter((visLayer) => isVisLayerWithError(visLayer));
+  if (!isEmpty(visLayersWithErrors)) {
+    // Aggregate by unique plugin resource type
+    const resourceTypes = [
+      ...new Set(visLayersWithErrors.map((visLayer) => visLayer.pluginResource.type)),
+    ];
+
+    let msgDetails = '';
+    resourceTypes.forEach((type, index) => {
+      const matchingVisLayers = visLayersWithErrors.filter(
+        (visLayer) => visLayer.pluginResource.type === type
+      );
+      if (index !== 0) msgDetails += '\n\n\n';
+      msgDetails += `-----${type}-----`;
+      matchingVisLayers.forEach((visLayer, idx) => {
+        if (idx !== 0) msgDetails += '\n';
+        msgDetails += `\nID: ${visLayer.pluginResource.id}`;
+        msgDetails += `\nMessage: "${visLayer.error?.message}"`;
+      });
+    });
+
+    const err = new Error(`Certain plugin resources failed to load on the ${visTitle} chart`);
+    // We set as the stack here so it can be parsed and shown cleanly in the details modal coming from the error toast notification.
+    err.stack = msgDetails;
+    return err;
+  } else {
+    return undefined;
+  }
+};
+
+/**
+ * Cleans up any stale saved objects caused by plugin resources being deleted. Kicks
+ * off an async call to delete the stale objs.
+ *
+ * @param augmentVisSavedObs the original augment-vis saved objs for this particular vis
+ * @param visLayers the produced VisLayers containing details if the resource has been deleted
+ * @param visualizationsLoader the visualizations saved object loader to handle deletion
+ */
+
+export const cleanupStaleObjects = (
+  augmentVisSavedObjs: ISavedAugmentVis[],
+  visLayers: VisLayer[],
+  loader: SavedAugmentVisLoader | undefined
+): void => {
+  const staleVisLayers = visLayers
+    .filter((visLayer) => isVisLayerWithError(visLayer))
+    .filter(
+      (visLayerWithError) => visLayerWithError.error?.type === VisLayerErrorTypes.RESOURCE_DELETED
+    );
+  if (!isEmpty(staleVisLayers)) {
+    const objIdsToDelete = [] as string[];
+    staleVisLayers.forEach((staleVisLayer) => {
+      // Match the VisLayer to its origin saved obj to extract the to-be-deleted saved obj ID
+      const deletedPluginResourceId = staleVisLayer.pluginResource.id;
+      const savedObjId = augmentVisSavedObjs.find(
+        (savedObj) => savedObj.pluginResource.id === deletedPluginResourceId
+      )?.id;
+      if (savedObjId !== undefined) objIdsToDelete.push(savedObjId);
```

Comment:
This would be faster if we throw the `augmentVisSavedObjs[].pluginResource.id` into a set or object than search an array for each.